### PR TITLE
Puncover will now unmangle cpp symbol names

### LINKED ISSUE #3: Improve C++ support
Would be nice to do some processing on C++ symbol names to make them look better.

This is what things look like right now with C++ (SDFat library):
screenshot from 2016-10-27 08 45 42

### PR DESCRIPTION
**Pull request by sarfata**
This requires an external tool (c++filt) which is called every time the
elf is reparsed with the list of all the symbols.

It is enabled all the time because it knows to not change C symbols.

This PR closes #3. 

Before:
screenshot from 2017-03-19 09 43 58

After:
screenshot from 2017-03-19 09 42 00

### CODE REVIEW

File: puncover/cppsupport.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import subprocess
+
+# See https://blog.flameeyes.eu/2010/06/c-name-demangling/ for context
+#
+# This solution courtesy of:
+# https://stackoverflow.com/questions/6526500/c-name-mangling-library-for-python/6526814
+# (slightly modified to return a dict instead of a list)
+def unmangle(names):
+    args = ['c++filt']
```
Comment (HBehrens) [inline puncover/cppsupport.py]: for the other binaries such as `objdump` the tool respects the toolchain dir that passed, e.g. https://github.com/HBehrens/puncover/blob/7896368183b93afd7db512c8cb29ddb0bab6fce6/puncover/collector.py#L261 Also, shouldn't this be `arm-none-eabi-c++filt`?

In the future, I would like to add another argument that helps distinguishing between `xyz` and `arm-none-eabi-xyz`. Centralizing the command line tooling already would help me doing so.

File: puncover/puncover.py [hunk 2]
```diff
@@ -37,6 +37,8 @@ def main():
                         help='location of your sources')
     parser.add_argument('--build_dir', dest='build_dir',
                         help='location of your build output')
+    parser.add_argument('--debug', action='store_true',
```
Comment (HBehrens) [inline puncover/puncover.py]: Thanks :) I should have done this way earlier :)

File: puncover/renderers.py [hunk 3]
```diff
@@ -25,6 +25,14 @@ def symbol_file(value):
     return value.get(collector.BASE_FILE, '__builtin')
 
 @jinja2.contextfilter
+def symbol_display_name_filter(context, value):
```
Comment (HBehrens) [inline puncover/renderers.py]: why do we need any of the filter code? in `unmangle_cpp_names()` you make sure that every symbol has a `.display_name` already.

File: puncover/templates/all_symbols.html.jinja [hunk 4]
```diff
@@ -15,7 +15,7 @@
         <tbody>
         {% for symbol in all_symbols %}
             <tr>
-            <td><a href="{{ symbol | symbol_url }}">{{ symbol.name | e}}</a></td>
+            <td><a href="{{ symbol | symbol_url }}">{{ symbol |symbol_display_name | e}}</a></td>
```
Comment (HBehrens) [inline puncover/templates/all_symbols.html.jinja]: here and in the other templates, can't you just write `symbol.display_name`?

**Comment by HBehrens** [comment]
Regarding my comment about lazy-initialization: I didn't realize that you are calling `c++filt` only once with the entire list. Maybe that's not so much of a bottleneck and initializing this lazily is not worth it all.

**Comment by sarfata** [comment]
I think I have addressed all your comments. Let me know what you think.

File: puncover/renderers.py [hunk 5]
```diff
@@ -93,18 +93,40 @@ def unique_filter(context, value):
...
+
+    # Get a clean display name - and a URL - for symbol names in comments
     #   b8:	f000 f8de 	bleq	278 &lt:__aeabi_dmul+0x1dc&gt:
     pattern = re.compile(r"&lt;(\w+)")
-    value = str(value)
     s = pattern.sub(lambda match: "&lt;"+linked_symbol_name(match.group(1)), value)
+
+    # Get a clean display name for symbol names in labels
+    # _ZN6Stream9readBytesEPcj():
+    # FIXME: Unfortunately symbols that have been inlined will not be in our global
```
Comment (HBehrens) [inline puncover/renderers.py]: I created #5 for this

**Pull request merged**